transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+    udp_mux: Arc<UDPMuxNewAddr>,
+
+    /// Future that drives reading from the UDP socket.
+    udp_mux_read_fut: Option<BoxFuture<'static, UDPMuxEvent>>,
+
+    /// `Keypair` identifying this peer
+    id_keys: identity::Keypair,
+
+    /// Set to `Some` if this [`Listener`] should close.
+    /// Optionally contains a [`TransportEvent::ListenerClosed`] that should be
+    /// reported before the listener's stream is terminated.
+    report_closed: Option<Option<<Self as Stream>::Item>>,
+}
+
+impl WebRTCListenStream {
+    /// Constructs a `WebRTCListenStream` for incoming connections.
+    fn new(
+        listener_id: ListenerId,
+        listen_addr: SocketAddr,
+        config: RTCConfiguration,
+        udp_mux: Arc<UDPMuxNewAddr>,
+        id_keys: identity::Keypair,
+    ) -> Self {
+        let in_addr = InAddr::new(listen_addr.ip());
+
+        WebRTCListenStream {
+            listener_id,
+            listen_addr,
+            in_addr,
+            config,
+            udp_mux,
+            udp_mux_read_fut: None,
+            id_keys,
+            report_closed: None,
+        }
+    }
+
+    /// Report the listener as closed in a [`TransportEvent::ListenerClosed`] and
+    /// terminate the stream.
+    fn close(&mut self, reason: Result<(), Error>) {
+        match self.report_closed {
+            Some(_) => debug!("Listener was already closed."),
+            None => {
+                // Report the listener event as closed.
+                let _ = self
+                    .report_closed
+                    .insert(Some(TransportEvent::ListenerClosed {
+                        listener_id: self.listener_id,
+                        reason,
+                    }));
+            }
+        }
+    }
+
+    /// Poll for a next If Event.
+    fn poll_if_addr(&mut self, cx: &mut Context<'_>) -> Option<<Self as Stream>::Item> {
+        match self.in_addr.poll_next_unpin(cx) {
+            Poll::Ready(mut item) => {
+                if let Some(item) = item.take() {
+                    // Consume all events for up/down interface changes.
+                    match item {
+                        Ok(IfEvent::Up(inet)) => {
+                            let ip = inet.addr();
+                            if self.listen_addr.is_ipv4() == ip.is_ipv4()
+                                || self.listen_addr.is_ipv6() == ip.is_ipv6()
+                            {
+                                let socket_addr = SocketAddr::new(ip, self.listen_addr.port());
+                                let ma = socketaddr_to_multiaddr(&socket_addr);
+                                debug!("New listen address: {}", ma);
+                                Some(TransportEvent::NewAddress {
+                                    listener_id: self.listener_id,
+                                    listen_addr: ma,
+                                })
+                            } else {
+                                self.poll_if_addr(cx)
```

Comment:
I would prefer not using recursion here. See rational in https://github.com/kpp/rust-libp2p/pull/19#discussion_r923989926.

If I understand correctly, the plan is to move this code into a place shared between the many transports, thus I am fine to leave it as is for now.

Reply:
> If I understand correctly, the plan is to move this code into a place shared between the many transports, thus I am fine to leave it as is for now.

That's the plan.